src/ray/object_manager/object_buffer_pool.cc

```diff
@@ -0,0 +1,235 @@
+#include "ray/object_manager/object_buffer_pool.h"
+
+namespace ray {
+
+ObjectBufferPool::ObjectBufferPool(const std::string &store_socket_name,
+                                   const uint64_t chunk_size) {
+  store_socket_name_ = store_socket_name;
+  chunk_size_ = chunk_size;
+}
+
+uint64_t ObjectBufferPool::GetNumChunks(uint64_t data_size) {
+  return ceil(static_cast<float>(data_size) / chunk_size_);
+}
+
+const ObjectBufferPool::ChunkInfo &ObjectBufferPool::GetChunk(const ObjectID &object_id,
+                                                              uint64_t data_size,
+                                                              uint64_t metadata_size,
+                                                              uint64_t chunk_index) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  if (get_failed_count_.count(object_id) != 0) {
+    // Ensure all gets fail if the first get fails.
+    get_failed_count_[object_id]--;
+    if (get_failed_count_[object_id] == 0) {
+      get_failed_count_.erase(object_id);
+    }
+    return errored_chunk_;
+  }
+  RAY_LOG(DEBUG) << "GetChunk " << object_id << " " << data_size << " " << metadata_size;
+  if (chunk_info_.count(object_id) == 0) {
+    std::shared_ptr<plasma::PlasmaClient> store_client = GetObjectStore();
+    plasma::ObjectBuffer object_buffer;
+    plasma::ObjectID plasma_id = ObjectID(object_id).to_plasma_id();
+    ARROW_CHECK_OK(store_client->Get(&plasma_id, 1, 0, &object_buffer));
+    if (object_buffer.data_size == -1) {
+      ReleaseObjectStore(store_client);
+      RAY_LOG(ERROR) << "Failed to get object";
+      // Ensure all gets fail if the first get fails.
+      get_failed_count_[object_id] = GetNumChunks(data_size) - 1;
+      return errored_chunk_;
+    }
+    RAY_CHECK(object_buffer.metadata->data() ==
+              object_buffer.data->data() + object_buffer.data_size);
+    RAY_CHECK(data_size == static_cast<uint64_t>(object_buffer.data_size +
+                                                 object_buffer.metadata_size));
+    auto *data = const_cast<uint8_t *>(object_buffer.data->data());
+    uint64_t num_chunks = BuildChunks(object_id, data, data_size, metadata_size);
+    buffer_state_.emplace(
```

Comment:
This seems to assume that if we do a `Get` on one chunk of an object, we are also going to call `Get` on every other chunk of the object. This also seems like it will break if we do simultaneous `Get`s on the same chunk of the same object. A less error-prone way of doing this might be to keep an actual reference count of the number of `Get` callers of each object. Then, this method would just increment the count, instead of setting the initial count to `num_chunks`.